src/core_modules/capture-core/metaDataMemoryStoreBuilders/programs/programsBuilder.js

```diff
@@ -185,21 +179,24 @@ async function getBuiltPrograms(
         locale,
     );
 
-    const promisePrograms = cachedPrograms.map(async (cachedProgram) => {
-        // $FlowFixMe[incompatible-call] automated comment
-        const program = await programFactory.build(cachedProgram);
-        return program;
-    });
-    const programs = await Promise.all(promisePrograms);
-    return programs;
+    const promisePrograms = cachedPrograms
+        .filter(cachedProgram => (
+            cachedProgram.trackedEntityTypeId &&
+            trackedEntityTypeCollection.get(cachedProgram.trackedEntityTypeId)
+        ))
```

Comment:
Here is basically what we discussed. Having it in the factory would result the class returning sometimes program as `undefined`. 

This is because 
```
    async build(cachedProgram: CachedProgram) {
        let program;
        if (cachedProgram.programType === 'WITHOUT_REGISTRATION') {
            program = new EventProgram((o) => {
                o.id = cachedProgram.id;
                o.access = cachedProgram.access;
                o.name = cachedProgram.displayName;
                o.shortName = cachedProgram.displayShortName;
                o.organisationUnits = cachedProgram.organisationUnits;
                o.categoryCombination = this._buildCategoryCombination(cachedProgram.categoryCombo);
            });
            const d2Stage = cachedProgram.programStages && cachedProgram.programStages[0];
            program.stage =
                await this.programStageFactory.build(
                    d2Stage,
                    program.id,
                );
        } else {
            program = new TrackerProgram((o) => {
                o.id = cachedProgram.id;
                o.access = cachedProgram.access;
                o.name = cachedProgram.displayName;
                o.shortName = cachedProgram.displayShortName;
                o.organisationUnits = cachedProgram.organisationUnits;
                // $FlowFixMe
                o.trackedEntityType = this.trackedEntityTypeCollection.get(cachedProgram.trackedEntityTypeId);
            });

            if (cachedProgram.programTrackedEntityAttributes) {
                program.searchGroups = await this.searchGroupFactory.build(
                    cachedProgram.programTrackedEntityAttributes,
                    cachedProgram.minAttributesRequiredToSearch,
                );

                // $FlowFixMe
                program.attributes = await this._buildProgramAttributes(cachedProgram.programTrackedEntityAttributes);
            }

            // $FlowFixMe
            await cachedProgram.programStages.asyncForEach(async (cachedProgramStage: CachedProgramStage) => {
                program.addStage(
                    await this.programStageFactory.build(
                        cachedProgramStage,
                        program.id,
                    ),
                );
            });

            program.enrollment = await this.enrollmentFactory.build(cachedProgram, program.searchGroups);
        }
        // $FlowFixMe
        program.icon = await ProgramFactory._buildProgramIcon(cachedProgram.style);

        return program;
    }

```

basically we would have to say return undefined when the program doesnt exist in `this.trackedEntityTypeCollection`. This wouldnt be very nice I believe. This is why I went with this alternative

Reply:
Yes, I concur with the approach 👍